File: Data Structures & Algorithms/level-order-traversal-of-binary-tree/submission-3.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    vector<vector<int>> res;
    vector<vector<int>> levelOrder(TreeNode* root) {
        dfs(root, 0);
        return res;
    }
    void dfs(TreeNode * root, int dep){
        if(!root) return;
        if(dep == res.size()){
            res.push_back({});
        }
        res[dep].push_back(root->val);
        dfs(root->left, dep+1);
        dfs(root->right, dep+1);
    }
};

File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        mp = {}
        for i in range(len(inorder)):
            mp[inorder[i]] = i
        i = 0
        def dfs(l,r):
            nonlocal i
            if i == len(preorder):
                return
            if l > r:
                return None
            val = preorder[i]
            idx = mp[val]
            node = TreeNode(val)
            
            i+=1
            node.left = dfs(l, idx-1)
            node.right = dfs(idx+1, r)
            return node
        return dfs(0, len(preorder)-1) 